In [1]:
from openai import OpenAI
from dotenv import load_dotenv
import os

load_dotenv()

api_key = os.getenv("OPENAI_API_KEY")
# print(f"API Key: {api_key}")

client = OpenAI()

completion = client.chat.completions.create(
  model="gpt-3.5-turbo",
  messages=[
    {"role": "system", "content": "You are a poetic assistant, skilled in explaining complex programming concepts with creative flair."},
    {"role": "user", "content": "Compose a poem that explains the concept of recursion in programming."}
  ]
)

In [2]:
print(completion.choices[0].message)

ChatCompletionMessage(content="In the realm of code, there lies a method divine,\nA looping dance that's elegant and oh so fine.\nRecursion, the art of self-calling grace,\nA technique that brings a smile to a programmer's face.\n\nLike a mirror reflecting its own image clear,\nRecursion calls itself, with no trace of fear.\nBreaking big problems into smaller bites,\nIt conquers complexity with dazzling lights.\n\nA function that calls itself, again and again,\nSolving puzzles with a recursive refrain.\nEach call delving deeper into the maze,\nUntil the goal is found, in a mesmerizing blaze.\n\nBut beware the infinite loop's cruel snare,\nCareful design is key, handle with care.\nBase cases like anchors in a stormy sea,\nGuide recursion safely, setting the code free.\n\nSo embrace this concept, let it be your friend,\nRecursion's power is yours to transcend.\nUnlocking mysteries, with each recursive call,\nIn the wondrous world of code, you'll stand tall.", role='assistant', function_c

In [3]:
print(completion.choices[0].message.content)

In the realm of code, there lies a method divine,
A looping dance that's elegant and oh so fine.
Recursion, the art of self-calling grace,
A technique that brings a smile to a programmer's face.

Like a mirror reflecting its own image clear,
Recursion calls itself, with no trace of fear.
Breaking big problems into smaller bites,
It conquers complexity with dazzling lights.

A function that calls itself, again and again,
Solving puzzles with a recursive refrain.
Each call delving deeper into the maze,
Until the goal is found, in a mesmerizing blaze.

But beware the infinite loop's cruel snare,
Careful design is key, handle with care.
Base cases like anchors in a stormy sea,
Guide recursion safely, setting the code free.

So embrace this concept, let it be your friend,
Recursion's power is yours to transcend.
Unlocking mysteries, with each recursive call,
In the wondrous world of code, you'll stand tall.


### READ DATASET FROM HUGGINGFACE

In [2]:
from datasets import load_dataset
import pandas as pd

/home/yasser/Documents/pa/.pa/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [3]:
df = load_dataset("gretelai/synthetic_text_to_sql", split='train')

In [4]:
df

Dataset({
    features: ['id', 'domain', 'domain_description', 'sql_complexity', 'sql_complexity_description', 'sql_task_type', 'sql_task_type_description', 'sql_prompt', 'sql_context', 'sql', 'sql_explanation'],
    num_rows: 100000
})

In [8]:
type(df)

datasets.arrow_dataset.Dataset

In [5]:
df[0]

{'id': 5097,
 'domain': 'forestry',
 'domain_description': 'Comprehensive data on sustainable forest management, timber production, wildlife habitat, and carbon sequestration in forestry.',
 'sql_complexity': 'single join',
 'sql_complexity_description': 'only one join (specify inner, outer, cross)',
 'sql_task_type': 'analytics and reporting',
 'sql_task_type_description': 'generating reports, dashboards, and analytical insights',
 'sql_prompt': 'What is the total volume of timber sold by each salesperson, sorted by salesperson?',
 'sql_context': "CREATE TABLE salesperson (salesperson_id INT, name TEXT, region TEXT); INSERT INTO salesperson (salesperson_id, name, region) VALUES (1, 'John Doe', 'North'), (2, 'Jane Smith', 'South'); CREATE TABLE timber_sales (sales_id INT, salesperson_id INT, volume REAL, sale_date DATE); INSERT INTO timber_sales (sales_id, salesperson_id, volume, sale_date) VALUES (1, 1, 120, '2021-01-01'), (2, 1, 150, '2021-02-01'), (3, 2, 180, '2021-01-01');",
 'sql'

## FORMAT THE DATA

In [ ]:
df = df.to_pandas()

In [35]:
df = df[['sql_prompt', 'sql']][:1000]
df

,sql_prompt,sql
0,What is the total volume of timber sold by eac...,"SELECT salesperson_id, name, SUM(volume) as to..."
1,List all the unique equipment types and their ...,"SELECT equipment_type, SUM(maintenance_frequen..."
2,How many marine species are found in the South...,SELECT COUNT(*) FROM marine_species WHERE loca...
3,What is the total trade value and average pric...,"SELECT trader_id, stock, SUM(price * quantity)..."
4,Find the energy efficiency upgrades with the h...,"SELECT type, cost FROM (SELECT type, cost, ROW..."
...,...,...
995,What is the total funding received by companie...,"SELECT company_id, SUM(amount) as total_fundin..."
996,How many publications were made by graduate st...,SELECT COUNT(Publications) FROM GraduateStuden...
997,Which space agencies have launched spacecraft ...,SELECT DISTINCT agency FROM spacecraft WHERE Y...
998,What's the total amount of climate finance com...,SELECT SUM(amount) FROM climate_finance WHERE ...


In [67]:
def convert_to_gpt35_format(dataset):
    fine_tuning_data = []
    for _, row in dataset.iterrows():
        fine_tuning_data.append({
            "messages": [
                {"role": "user", "content": row['sql_prompt']},
                {"role": "assistant", "content": row['sql']},
            ]
        })
    return fine_tuning_data

In [68]:
converted_data = convert_to_gpt35_format(df)

In [70]:
converted_data[0]

{'messages': [{'role': 'user',
   'content': 'What is the total volume of timber sold by each salesperson, sorted by salesperson?'},
  {'role': 'assistant',
   'content': 'SELECT salesperson_id, name, SUM(volume) as total_volume FROM timber_sales JOIN salesperson ON timber_sales.salesperson_id = salesperson.salesperson_id GROUP BY salesperson_id, name ORDER BY total_volume DESC;'}]}

## Creating Training and Validation Sets

In [71]:
from sklearn.model_selection import train_test_split

# Stratified splitting. Assuming 'Top Category' can be used for stratification
train_data, val_data = train_test_split(
    converted_data,
    test_size=0.2,
    random_state=42  # for reproducibility
)

In [72]:
train_data

[{'messages': [{'role': 'user',
    'content': 'How many companies were founded by women in the San Francisco Bay Area?'},
   {'role': 'assistant',
    'content': "SELECT COUNT(*) FROM companies WHERE city='San Francisco' AND state='CA' AND founder_gender='female';"}]},
 {'messages': [{'role': 'user',
    'content': 'List all chemical batches with a pH level above 9, produced in the USA since January 2021, along with their production dates.'},
   {'role': 'assistant',
    'content': "SELECT id, pH, production_date FROM ChemicalBatches WHERE pH > 9 AND production_date >= '2021-01-01';"}]},
 {'messages': [{'role': 'user',
    'content': 'What is the minimum ocean acidification level recorded in the Pacific Ocean, and which research station had this level?'},
   {'role': 'assistant',
    'content': "SELECT research_station.station_name, oa.level AS min_level FROM ocean_acidification oa JOIN (SELECT location, MIN(level) AS min_level FROM ocean_acidification WHERE region = 'Pacific Ocean' G

In [73]:
val_data[0]

{'messages': [{'role': 'user',
   'content': 'What is the minimum price of vegan skincare products sold in the United Kingdom?'},
  {'role': 'assistant',
   'content': "SELECT MIN(price) FROM SkincareProducts WHERE is_vegan = TRUE AND country = 'United Kingdom';"}]}

## Creating JSONL Files

In [74]:
import json

def write_to_jsonl(data, file_path):
    with open(file_path, 'w') as file:
        for entry in data:
            json.dump(entry, file)
            file.write('\n')

training_file_name = "train.jsonl"
validation_file_name = "val.jsonl"

write_to_jsonl(train_data, training_file_name)
write_to_jsonl(val_data, validation_file_name)

## Uploading Data and Starting the Fine-Tuning Job

In [75]:
from openai import OpenAI
from dotenv import load_dotenv
import os

load_dotenv()

api_key = os.getenv("OPENAI_API_KEY")

client = OpenAI()


# Upload Training and Validation Files
training_file = client.files.create(
    file=open(training_file_name, "rb"), purpose="fine-tune"
)
validation_file = client.files.create(
    file=open(validation_file_name, "rb"), purpose="fine-tune"
)

In [76]:
print("Training_id : ", training_file.id)
print("Validation id :", validation_file.id)

Training_id :  file-5rlIi2wCQNzMuzVur6FjMdpg
Validation id : file-sEdOw55HvHQzNpCePFrNRt9K


## Create Fine-Tuning Job

In [77]:
response = client.fine_tuning.jobs.create(
    training_file=training_file.id,
    validation_file=validation_file.id,
    model="gpt-3.5-turbo",
    suffix="text-to-sql",
)

In [78]:
response

FineTuningJob(id='ftjob-wL67oqMDw1ovE2frIyWK86CZ', created_at=1715435567, error=Error(code=None, message=None, param=None), fine_tuned_model=None, finished_at=None, hyperparameters=Hyperparameters(n_epochs='auto', batch_size='auto', learning_rate_multiplier='auto'), model='gpt-3.5-turbo-0125', object='fine_tuning.job', organization_id='org-jEJmvnn0sAe0ifDnYRWxdmRo', result_files=[], status='validating_files', trained_tokens=None, training_file='file-5rlIi2wCQNzMuzVur6FjMdpg', validation_file='file-sEdOw55HvHQzNpCePFrNRt9K', user_provided_suffix='text-to-sql', seed=645630885, estimated_finish=None, integrations=[])

In [90]:
response.id

'ftjob-wL67oqMDw1ovE2frIyWK86CZ'

In [98]:
client.fine_tuning.jobs.retrieve(response.id)

FineTuningJob(id='ftjob-wL67oqMDw1ovE2frIyWK86CZ', created_at=1715435567, error=Error(code=None, message=None, param=None), fine_tuned_model='ft:gpt-3.5-turbo-0125:personal:text-to-sql:9NiqfkTN', finished_at=1715440968, hyperparameters=Hyperparameters(n_epochs=3, batch_size=1, learning_rate_multiplier=2), model='gpt-3.5-turbo-0125', object='fine_tuning.job', organization_id='org-jEJmvnn0sAe0ifDnYRWxdmRo', result_files=['file-qTFxcOvRh7bMwMoxIrTp81Iv'], status='succeeded', trained_tokens=135678, training_file='file-5rlIi2wCQNzMuzVur6FjMdpg', validation_file='file-sEdOw55HvHQzNpCePFrNRt9K', user_provided_suffix='text-to-sql', seed=645630885, estimated_finish=None, integrations=[])

In [19]:
client.fine_tuning.jobs.list(limit=1)

SyncCursorPage[FineTuningJob](data=[FineTuningJob(id='ftjob-wL67oqMDw1ovE2frIyWK86CZ', created_at=1715435567, error=Error(code=None, message=None, param=None), fine_tuned_model='ft:gpt-3.5-turbo-0125:personal:text-to-sql:9NiqfkTN', finished_at=1715440968, hyperparameters=Hyperparameters(n_epochs=3, batch_size=1, learning_rate_multiplier=2), model='gpt-3.5-turbo-0125', object='fine_tuning.job', organization_id='org-jEJmvnn0sAe0ifDnYRWxdmRo', result_files=['file-qTFxcOvRh7bMwMoxIrTp81Iv'], status='succeeded', trained_tokens=135678, training_file='file-5rlIi2wCQNzMuzVur6FjMdpg', validation_file='file-sEdOw55HvHQzNpCePFrNRt9K', user_provided_suffix='text-to-sql', seed=645630885, estimated_finish=None, integrations=[])], object='list', has_more=True)

SyncCursorPage[FineTuningJob]
(data=[FineTuningJob(id='ftjob-wL67oqMDw1ovE2frIyWK86CZ',
created_at=1715435567, error=Error(code=None, message=None, param=None),
fine_tuned_model='ft:gpt-3.5-turbo-0125:personal:text-to-sql:9NiqfkTN', 
finished_at=1715440968, 

hyperparameters=Hyperparameters(n_epochs=3, batch_size=1, learning_rate_multiplier=2), 

model='gpt-3.5-turbo-0125', object='fine_tuning.job', organization_id='org-jEJmvnn0sAe0ifDnYRWxdmRo', 
result_files=['file-qTFxcOvRh7bMwMoxIrTp81Iv'], 
status='succeeded', trained_tokens=135678, training_file='file-5rlIi2wCQNzMuzVur6FjMdpg', validation_file='file-sEdOw55HvHQzNpCePFrNRt9K', 
user_provided_suffix='text-to-sql', seed=645630885, estimated_finish=None, integrations=[])], object='list', has_more=True)

### TEST FINETUNED MODEL

### id : ft:gpt-3.5-turbo-0125:personal:text-to-sql:9NiqfkTN

In [1]:
from openai import OpenAI
from dotenv import load_dotenv
from datasets import load_dataset
import pandas as pd
import os

/home/yasser/Documents/pa/.pa/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [3]:
df = load_dataset("gretelai/synthetic_text_to_sql", split='train')
df = df.to_pandas()

In [23]:
df.head(5)

,id,domain,domain_description,sql_complexity,sql_complexity_description,sql_task_type,sql_task_type_description,sql_prompt,sql_context,sql,sql_explanation
0,5097,forestry,Comprehensive data on sustainable forest manag...,single join,"only one join (specify inner, outer, cross)",analytics and reporting,"generating reports, dashboards, and analytical...",What is the total volume of timber sold by eac...,"CREATE TABLE salesperson (salesperson_id INT, ...","SELECT salesperson_id, name, SUM(volume) as to...","Joins timber_sales and salesperson tables, gro..."
1,5098,defense industry,"Defense contract data, military equipment main...",aggregation,"aggregation functions (COUNT, SUM, AVG, MIN, M...",analytics and reporting,"generating reports, dashboards, and analytical...",List all the unique equipment types and their ...,CREATE TABLE equipment_maintenance (equipment_...,"SELECT equipment_type, SUM(maintenance_frequen...",This query groups the equipment_maintenance ta...
2,5099,marine biology,"Comprehensive data on marine species, oceanogr...",basic SQL,basic SQL with a simple select statement,analytics and reporting,"generating reports, dashboards, and analytical...",How many marine species are found in the South...,"CREATE TABLE marine_species (name VARCHAR(50),...",SELECT COUNT(*) FROM marine_species WHERE loca...,This query counts the number of marine species...
3,5100,financial services,Detailed financial data including investment s...,aggregation,"aggregation functions (COUNT, SUM, AVG, MIN, M...",analytics and reporting,"generating reports, dashboards, and analytical...",What is the total trade value and average pric...,"CREATE TABLE trade_history (id INT, trader_id ...","SELECT trader_id, stock, SUM(price * quantity)...",This query calculates the total trade value an...
4,5101,energy,Energy market data covering renewable energy s...,window functions,"window functions (e.g., ROW_NUMBER, LEAD, LAG,...",analytics and reporting,"generating reports, dashboards, and analytical...",Find the energy efficiency upgrades with the h...,"CREATE TABLE upgrades (id INT, cost FLOAT, typ...","SELECT type, cost FROM (SELECT type, cost, ROW...",The SQL query uses the ROW_NUMBER function to ...


In [10]:
load_dotenv()

api_key = os.getenv("OPENAI_API_KEY")

In [11]:
print(f"API Key: {api_key}")

API Key: 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


In [27]:
df[df['id'] == 5099]['sql_prompt'][2]

'How many marine species are found in the Southern Ocean?'

In [28]:
client = OpenAI()

completion = client.chat.completions.create(
  model="ft:gpt-3.5-turbo-0125:personal:text-to-sql:9NiqfkTN",
  messages=[
    {"role": "user", 
     "content": "How many marine species are found in the Southern Ocean?"
     }
  ]
)

In [29]:
print(completion.choices[0].message.content)

SELECT COUNT(*) FROM marine_species WHERE location = 'Southern Ocean';


In [30]:
df[df['id'] == 5099]['sql'][2]

"SELECT COUNT(*) FROM marine_species WHERE location = 'Southern Ocean';"